contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.20;
+
+import "forge-std/console2.sol";
+import "../libraries/SwapIntention.sol";
+import {IQuoter} from "../interfaces/IQuoter.sol";
+import {PoolTicksCounter} from "../libraries/PoolTicksCounter.sol";
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {SafeCast} from "@uniswap/v4-core/contracts/libraries/SafeCast.sol";
+import {Pool} from "@uniswap/v4-core/contracts/libraries/Pool.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {PoolManager} from "@uniswap/v4-core/contracts/PoolManager.sol";
+
+contract Quoter is IQuoter {
+    using PoolIdLibrary for PoolKey;
+    using SafeCast for *;
+    using Hooks for IHooks;
+
+    // v4 Singleton contract
+    IPoolManager poolManager;
+
+    constructor(address _poolManager) {
+        poolManager = IPoolManager(_poolManager);
+    }
+
+    function fillSlot0(PoolId id) private view returns (Pool.Slot0 memory slot0) {
+        //TODO: extsload when storage is stable?
+        (slot0.sqrtPriceX96, slot0.tick,,) = poolManager.getSlot0(id);
+
+        return slot0;
+    }
+
+    function validateRevertReason(bytes memory reason) private pure returns (bytes memory) {
+        if (reason.length < 96) {
```

Comment:
also seems like this validation should differ between exactInput and exactInputSingle since they have different sizes of return values, no? i would assume we need separate validation for each

Reply:
oh i guess 96 is from exactInputSingle revert which is 3 words long, yea comment would be good :D